### PR TITLE
Adding picocli codestart

### LINKED ISSUE #12633: Create the Picocli Example codestart in Java & Kotlin
**Description**
Create the Picocli Example codestart in Java & Kotlin

as part of epic: https://github.com/quarkusio/quarkus/issues/8134

### Contributing an example Codestart

- doc: https://github.com/quarkusio/quarkus/blob/master/independent-projects/tools/codestarts/README.adoc
- codestarts dir: https://github.com/quarkusio/quarkus/tree/master/devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/examples

### Codestart Core

- code: https://github.com/quarkusio/quarkus/tree/master/independent-projects/tools/codestarts
- core codestarts dir: https://github.com/quarkusio/quarkus/tree/master/devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/core

### PR DESCRIPTION
**Pull request by jtama**
Fixes #12633

### CODE REVIEW

File: devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/picocli-example/base/README.tpl.qute.md
Comment (ia3andy) [inline devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/picocli-example/base/README.tpl.qute.md:5]:
````suggestion
Picocli example

Hello and goodbye are civilization fundamentals. Let's not forget it with this picocli application by changing the `command` and `parameters`.

Also for picocli applications the dev mode is supported. When running dev mode, the picocli application is executed and on press of the Enter key, is restarted.

As command mode applications will often require arguments to be passed on the commandline, this is also possible in dev mode via:
```shell script
{buildtool.cli} {buildtool.cmd.dev} {#if buildtool.cli == 'gradle'}--quarkus-args{#else}-Dquarkus.args={/if}='help'
```

Guide: https://quarkus.io/guides/picocli
````

File: devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/picocli-example/codestart.yml
Comment (ia3andy) [inline devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/picocli-example/codestart.yml:6]:
```suggestion
tags: example
```

File: devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/picocli-example/codestart.yml
Comment (ia3andy) [inline devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/picocli-example/codestart.yml:25]:
```suggestion
```

You can remove that part, it was used because the title and description is also used in the html for the cloud codestarts.

File: devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/picocli-example/java/src/main/java/org/acme/picocli/GoodbyeCommand.java
Comment (ia3andy) [inline devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/picocli-example/java/src/main/java/org/acme/picocli/GoodbyeCommand.java:12]:
```suggestion
    @Option(names = {"--times", "-t"}, defaultValue = "1", description = "How many time should we say goodbye")
```

File: devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/picocli-example/kotlin/src/main/java/org/acme/picocli/GoodbyeCommand.kt
Comment (ia3andy) [inline devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/picocli-example/kotlin/src/main/java/org/acme/picocli/GoodbyeCommand.kt]:
```suggestion
    @CommandLine.Option(names = ["--times", "-t"], defaultValue = "1", description = ["How many time should we say goodbye"])
```

**Comment by jtama-op** [comment]
@ia3andy Actually there's a few thing I don't understand.
The `quarkus-platform-descriptor-json` tests generate the codestart in the exemple folder. And the unit tests I added in the `integration-tests-devtools` module generates the a kotlin project with the good classes but in a java directory structure...

**Review by ia3andy — CHANGES_REQUESTED** [review]
> I think this exemple should be a singleton-exemple. What do you think ? Waiting for your input I added dedicated unit tests, and excluded it from the IT.

I think normal example is fine, One app should be able to do REST and Command mode :), please remove the tag as suggested and move the directory.

> The code isn't tested. In fact it already bothers me for my project, there aren't any testing samples. How do I fake command agrs in a test env...

Yeah it's really missing, we have a feature request already: https://github.com/quarkusio/quarkus/issues/10212

> The singleton-exemple helped !

Are the tests failing is you don't do this?

> Finally, How do you mark a PR as draft ?

You should have a button in the PR to convert it to a draft

**Comment by ia3andy** [comment]
Thanks a lot for this contribution :) !!!!

File: devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/picocli-example/codestart.yml
Comment (jtama-op) [reply devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/picocli-example/codestart.yml:25]: Which is not the case anymore ?

Comment (ia3andy) [reply devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/singleton-examples/picocli-example/codestart.yml:25]: It is, but not for this example which doesn't include any web feature.

**Comment by ia3andy** [comment]
> @ia3andy Actually there's a few thing I don't understand.
> The `quarkus-platform-descriptor-json` tests generate the codestart in the exemple folder. And the unit tests I added in the `integration-tests-devtools` module generates the a kotlin project with the good classes but in a java directory structure...

I don't really get what you mean here, could you come on zulip to discuss this? https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Codestarts.2E.2E.2E

File: integration-tests/devtools/src/test/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartGenerationTest.java
Comment (ia3andy) [inline integration-tests/devtools/src/test/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartGenerationTest.java:340]: You don't need to add the codestart, it will be autoselected with the extension, but first you need to add it to the extension metadata in this file:
https://github.com/quarkusio/quarkus/blob/master/extensions/picocli/runtime/src/main/resources/META-INF/quarkus-extension.yaml#L10

```json
---
name: "Picocli"
metadata:
  keywords:
  - "picocli"
  - "command-line"
  guide: "https://quarkus.io/guides/picocli"
  categories:
  - "miscellaneous"
  status: "experimental"
  codestart: picocli

```

File: integration-tests/devtools/src/test/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartGenerationTest.java
Comment (ia3andy) [inline integration-tests/devtools/src/test/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartGenerationTest.java:318]: Same here

**Review by ia3andy — CHANGES_REQUESTED** [review]
Also please rebase all to one commit :)

File: integration-tests/devtools/src/test/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartGenerationTest.java
Comment (jtama-op) [reply integration-tests/devtools/src/test/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartGenerationTest.java:340]: Ok, this one is missing from guide I think.

Comment (ia3andy) [reply integration-tests/devtools/src/test/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartGenerationTest.java]: @jtama-op https://github.com/ia3andy/quarkus/blob/cs-doc/independent-projects/tools/codestarts/README.adoc#writing-a-codestart-with-example-code-for-an-extension

Look at the When done part :)

File: devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/examples/picocli-example/java/src/main/java/org/acme/picocli/HelloCommand.java
Comment (ia3andy) [inline devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/examples/picocli-example/java/src/main/java/org/acme/picocli/HelloCommand.java:17]: this felt a bit weird to have them exclusive, maybe `--firt-name` and `--nick-name`? would make more sense?

Comment (jtama-op) [reply devtools/platform-descriptor-json/src/main/resources/codestarts/quarkus/examples/picocli-example/java/src/main/java/org/acme/picocli/HelloCommand.java:17]: Ok yes of course! I only intended to show the exclusive feature, I will make the change.

**Comment by jtama-op** [comment]
@ia3andy Changes made, commits squashed, CI ok.

**Comment by ia3andy** [comment]
@gastaldi would you have a look to that one? I think it would be better to have another set of eyes on it :). Who is responsible for the picocli extension? We should maybe also add them..

**Comment by gastaldi** [comment]
I think @mgorniew and @remkop would be interested in this

**Comment by remkop** [comment]
@gastaldi Thank you for the notification, and @jtama-op thank you for doing this!
I will take a detailed look this weekend.

**Comment by jtama-op** [comment]
Thanks for the extension work !

**Comment by geoand** [comment]
Thanks for this work!

Is this ready to be merged?

**Comment by ia3andy** [comment]
@geoand I was waiting for a "business" review by the extension owners, on the codestart part it's good.

**Comment by remkop** [comment]
Looks good to me, very nice to see this! 👍

**Review by ia3andy — APPROVED** [review]

**Pull request merged**
